### PR TITLE
GBS: Add ability to simulate losses in GBS

### PR DESCRIPTION
**Pull request by trbromley**
Since realistic GBS devices will have losses, it makes sense to provide this in the applications layer. Since loss is already part of SF, we only need to change the program fed to the SF Engine, as well as updating tests and docs.

### CODE REVIEW

File: strawberryfields/gbs/sample.py [hunk 1]
```diff
@@ -142,6 +151,10 @@ def sample(A: np.ndarray, n_mean: float, n_samples: int = 1, threshold: bool = T
     with p.context as q:
         sf.ops.GraphEmbed(A, mean_photon_per_mode=mean_photon_per_mode) | q
 
+        if loss:
```
Comment (nquesada) [inline strawberryfields/gbs/sample.py]: seems like loss is a Boolean here but a float elsewhere.

Comment (nquesada) [reply strawberryfields/gbs/sample.py]: By the way, applying loss in the Gaussian backend is very cheap so you loose very little by always applying it.

Comment (ixfoduap) [reply strawberryfields/gbs/sample.py]: Thanks Nico! Actually, `if` evaluates to `True` for any non-zero float, and is `False` if the float is zero

File: strawberryfields/gbs/similarity.py [hunk 2]
```diff
@@ -345,6 +351,7 @@ def prob_orbit_mc(graph: nx.Graph, orbit: list, n_mean: float = 5, samples: int
         orbit (list[int]): orbit for which to estimate the probability
         n_mean (float): total mean photon number of the GBS device
         samples (int): number of samples used in the Monte Carlo estimation
+        loss (float): fraction of photons lost in GBS
```
Comment (ixfoduap) [inline strawberryfields/gbs/similarity.py]: This function probably needs to raise an error if the loss is not between 0 and 1

File: strawberryfields/gbs/similarity.py [hunk 3]
```diff
@@ -400,6 +413,7 @@ def prob_event_mc(
         max_count_per_mode (int): maximum number of photons per mode in the event
         n_mean (float): total mean photon number of the GBS device
         samples (int): number of samples used in the Monte Carlo estimation
+        loss (float): fraction of photons lost in GBS
```
Comment (ixfoduap) [inline strawberryfields/gbs/similarity.py]: Same here, function probably needs to raise an error if the loss is not between 0 and 1

File: strawberryfields/gbs/similarity.py [hunk 4]
```diff
@@ -509,6 +529,6 @@ def feature_vector_mc(
         raise ValueError("Maximum number of photons per mode must be at least one")
```
Comment (ixfoduap) [inline strawberryfields/gbs/similarity.py]: Same here, raise an error if the loss is not between 0 and 1

File: tests/gbs/test_sample.py [hunk 5]
```diff
@@ -74,6 +80,46 @@ def test_pnr(self, monkeypatch, adj):
...
+        mock_eng_run = mock.MagicMock()
+
+        with monkeypatch.context() as m:
+            m.setattr(sf.Engine, "run", mock_eng_run)
+            sample.sample(A=adj, n_mean=1, threshold=False)
+            p_func = mock_eng_run.call_args[0][0]
+
+        assert not all([isinstance(op, sf.ops.LossChannel) for op in p_func.circuit])
+
+    def test_all_loss(self, monkeypatch, adj, dim):
```
Comment (ixfoduap) [inline tests/gbs/test_sample.py]: Great test!

File: strawberryfields/gbs/similarity.py (hunk 2 above)
Comment (trbromley) [reply strawberryfields/gbs/similarity.py]: This has been added in: https://github.com/XanaduAI/strawberryfields/pull/223/commits/0035650994a5feb5f643201f54ce9eee4c913885, along with some other tests that were missing. I feel we could probably get away with having fewer tests though.

**Comment by trbromley** [comment]
Thanks @ixfoduap @nquesada.
Should be ready for another look!

File: strawberryfields/gbs/sample.py (hunk 1 above)
Comment (nquesada) [reply strawberryfields/gbs/sample.py]: I always though Python got Booleans precisely to not do that, like in C. Anyway, if that is how things work, then it's fine.

**Comment by josh146** [comment]
Note that we plan on adding hardware simulators in the future that have 'built in' loss, so when this is live, the `LossChannel` gates will need to be ignored if these simulators are used.

**Comment by josh146** [comment]
Thinking about this a bit more: the best solution will be to return the use of `backend_options={}` as an argument at some point in the future. This will allow the user to dispatch to a backend of their choosing (remote/local/gaussian/hardware/chip0sim), while also packaging up any options that are exclusive to that backend (loss, cutoff, etc).

**Comment by trbromley** [comment]
Thanks @josh146, agreed! For now we can use this but can definitely update as the lower layers of SF evolve.

**Pull request merged**